[ckclion] iP

ckclion opened the pull request:
### Ratchet

> "My time has come." – Master Oogway ([Source](https://www.imdb.com/title/tt0441773/characters/nm0453641))

Ratchet remembers your tasks for you. It's,

- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/ckclion/ip/releases/tag/A-Jar)
2. double-click it
3. add your tasks
4. let it manage your tasks for you :partying_face:

And it is **FREE**!

Features:

- [x] Managing `todos`
- [x] Managing `deadlines`
- [x] Managing `events`
- [x] GUI
##
All of it starts from the `main` method:
```java
public static void main(String[] args) {
    new Ratchet().run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ckclion committed:
Update Duke.java

* Renamed Duke.java to Ratchet.java
* Added greet function
* Added exit function

ckclion committed:
Add echo functionality to Ratchet.java

ckclion committed:
Add list functionality to Ratchet.java

ckclion committed:
Add mark and unmark functionality to tasks

ckclion committed:
Add support for tracking three types of tasks

ckclion committed:
Add Automated Text UI Testing

ckclion committed:
Add minimal error handling using exceptions

ckclion committed:
Add delete functionality to Ratchet.java

ckclion committed:
Add load data functionality

ckclion committed:
Add save data functionality

ckclion committed:
Merge branch 'branch-Level-7'

ckclion committed:
Add date to tasks

ckclion committed:
Merge branch 'branch-Level-8'

ckclion committed:
Refactor code by extracting out more classes

ckclion committed:
Merge branch 'branch-A-MoreOOP'

ckclion committed:
Refactor classes into packages

ckclion committed:
Merge branch 'branch-A-Packages'

ckclion committed:
Merge branch 'add-gradle-support'

ckclion committed:
Add gradle support

ckclion committed:
Add JUnit tests with minimal test cases

ckclion committed:
Update archiveBaseName in build.gradle

ckclion committed:
Add find task functionality

ckclion committed:
Refactor code to comply with coding standards

ckclion committed:
Add JavaDoc to some classes and methods

ckclion committed:
Merge branch 'branch-A-JavaDoc'

ckclion committed:
Merge branch 'branch-A-CodingStandard'

ckclion committed:
Merge branch 'branch-Level-9'

ckclion committed:
Add JavaDoc to all public classes and methods

ckclion committed:
Add checkstyle

ckclion committed:
Merge branch 'branch-A-CheckStyle'

ckclion committed:
Update text-ui to gui

ckclion committed:
Merge branch 'branch-Level-10'

hogretup commented on src/main/java/ratchet/Main.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * A GUI for Duke using FXML.
+ */
+public class Main extends Application {
+
+    private final Ratchet ratchet = new Ratchet();
+
+    @Override
+    public void start(Stage stage) {
+        try {
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
+            AnchorPane ap = fxmlLoader.load();
+            Scene scene = new Scene(ap);
+            stage.setScene(scene);
+            fxmlLoader.<MainWindow>getController().setRatchet(ratchet);
+            stage.show();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+}
```
very clean and readable code

Nihirraa commented on src/main/java/ratchet/parser/Parser.java:
```diff
@@ -0,0 +1,143 @@
+package ratchet.parser;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeParseException;
+
+import ratchet.command.AddCommand;
+import ratchet.command.Command;
+import ratchet.command.DeleteCommand;
+import ratchet.command.ExitCommand;
+import ratchet.command.FindCommand;
+import ratchet.command.ListCommand;
+import ratchet.command.MarkCommand;
+import ratchet.command.UnmarkCommand;
+import ratchet.exception.InvalidCommandArgumentException;
+import ratchet.exception.InvalidCommandException;
+import ratchet.exception.RatchetException;
+import ratchet.task.DeadlineTask;
+import ratchet.task.EventTask;
+import ratchet.task.Task;
+import ratchet.task.TodoTask;
```
I like how the import statements have been written following consistency.

Nihirraa commented on src/main/java/ratchet/storage/Storage.java:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    private void initFile() {
+        try {
+            new File(PATH_TO_DIRECTORY).mkdir();
+            new File(PATH_TO_TASK_FILE).createNewFile();
+        } catch (IOException e) {
+            System.out.println(e.getMessage());
+            System.exit(-1);
+        }
+    }
+}
```
include Javadoc comments for methods

Nihirraa commented on src/main/java/ratchet/storage/Storage.java:
```diff
@@ -0,0 +1,80 @@
+ * Class to handle the storage of the bot.
+ */
+public class Storage {
+    private static final String PATH_TO_DIRECTORY = "data";
+    private static final String PATH_TO_TASK_FILE = "data/ratchet.txt";
```
good and apt naming of variables

Nihirraa commented on src/main/java/ratchet/task/EventTask.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + from.format(DateTimeFormatter.ofPattern("MMM d yyyy"))
+                + " to: " + to.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
can use a variable instead of hardcoding, like:
``` suggested
private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("MMM d yyyy");
```

Nihirraa reviewed (commented):
Overall a very clear, understandable and well written code!

ckclion committed:
Add assertions

Some logic in code are under assumptions and these assumptions are not
being checked.

If there are new logic added that does not account for these
assumptions, the application will break.

Let's add assertions to account for these assumptions.

ckclion committed:
Refactor: Improve code quality

Current code is not very good as it violates some SLAP principles as
well as following some error-prone practices.

Improving code quality will not only help make bug fixing in the future
easier but it will also help maintainability.

Let's,
* refactor functions so that they are following slap
* add default cases to if else statements to make it safer

ckclion committed:
Merge pull request #1 from ckclion/branch-A-Assertions

Add assertions

ckclion committed:
Merge branch 'master' into branch-A-CodeQuality

ckclion committed:
Merge pull request #2 from ckclion/branch-A-CodeQuality

 Refactor: Improve code quality

ckclion committed:
Add mass operation to delete, mark, unmark

ckclion committed:
Merge branch 'branch-C-MassOps'

ckclion committed:
Update GUI

ckclion committed:
Merge branch 'branch-A-BetterGui'

ckclion committed:
Add error handling for empty index

ckclion committed:
Merge branch 'branch-A-MoreErrorHandling'

ckclion committed:
Update user guide to docs\README.md

ckclion committed:
Add exception to handle index not in list

ckclion committed:
Add more JUnit tests for parser

ckclion committed:
Add more JavaDoc comments

ckclion committed:
Update docs/README.md

ckclion committed:
Update README.md to include credits to images

ckclion committed:
Update README.md to fix formatting issues

ckclion committed:
Update README.md to fix formatting issues in user guide
